### ISSUE TITLE
java.lang.NoClassDefFoundError: org/apache/avro/generic/GenericRecord

### ISSUE THREAD

**Issue by lionell88** [lensesio/stream-reactor#821]
# Issue Guidelines

Please review these questions before submitting any issue?

### Kafka-connect-mqtt

When your connector reads a packet from mqtt it doesn't forward it to Kafka topic and I got this exception:

MqttException (0) - java.lang.NoClassDefFoundError: org/apache/avro/generic/GenericRecord
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:228)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: java.lang.NoClassDefFoundError: org/apache/avro/generic/GenericRecord
	at com.datamountaineer.streamreactor.common.converters.MsgKey$.<init>(MsgKey.scala:26)
	at com.datamountaineer.streamreactor.common.converters.MsgKey$.<clinit>(MsgKey.scala)
	at com.datamountaineer.streamreactor.connect.converters.source.BytesConverter.convert(BytesConverter.scala:37)
	at com.datamountaineer.streamreactor.connect.mqtt.source.MqttManager.messageArrived(MqttManager.scala:109)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.deliverMessage(CommsCallback.java:519)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.handleMessage(CommsCallback.java:417)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:214)
	... 1 more
Caused by: java.lang.ClassNotFoundException: org.apache.avro.generic.GenericRecord
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:436)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
	at org.apache.kafka.connect.runtime.isolation.PluginClassLoader.loadClass(PluginClassLoader.java:103)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 8 more

So it seems to disconnect because doesn't find  that lib. 

In the connect-distribuited.properties I configured the path to the full Stream-reactor lib.

![image](https://user-images.githubusercontent.com/2744266/146367247-9898a8c2-a934-46ad-a4a0-8ad135dbf87c.png)

Thankyou for your precious interest.

**Comment by lionell88** [comment]
when each connector start, it seems not be able to load any of the dependencies classes in the jars

![image](https://user-images.githubusercontent.com/2744266/146376754-e2e691d8-e03a-4caa-8269-c2d153e5a34b.png)

**Comment by davidsloan** [comment]
Hi there, thanks for raising this issue.

Please try adjusting your plugin path to point to the full existing Kafka Connect plugins folder.  There should be some jars in there already that are necessary.

The plugins directory should be laid out like this

```
--plugins
  |___ plugin1
  |___ plugin2
  |___ stream reactor
```
Stream reactor shouldn't contain the lib directory.

Also it is best to just copy in the connectors you need.  You shouldn't need the kafka-connect-common jars in there as all required dependencies are packaged inside the individual connector jars.

**Comment by lionell88** [comment]
Thankyou for you reply. I got this:
java.lang.NoClassDefFoundError: io/confluent/connect/avro/AvroData

Then I unzipped the Confluent  jar libs into the plugin path. I pointed to the root plugin path set in the connect-distribuited. properties but no way to load that library: always the same exception. So I put them into the kafka libs folder (except for your mqtt connector jar) and now it works. I don't think this is the best approach but it was the only one for me.
